### DISCUSSION TITLE
Offline sessions lazy loaded by default

### DISCUSSION THREAD

**Discussion by martin-kanis** [keycloak/keycloak#9619] [Ideas]
In https://github.com/keycloak/keycloak/pull/9414 I switched the default behavior regarding offline sessions preloading. An offline session is lazy loaded from a database when requested. The previous behavior where offline sessions were preloaded from the database caused problems with slow server startup in deployments with many offline sessions. 

### Required changes
To keep Keycloak working with lazy loading of offline sessions I needed to change several things that are described below.

#### Search offline sessions by broker session ID
We can obtain stream of offline sessions  filtered by `userId`, `brokerUserId` or `brokerSessionId` in current code base. However, the `PersistentUserSessionEntity` has only `userId` as a property (column) from those three fields. The rest is encoded as JSON in data property (column). That means, we need to figure out how to search `PersistentUserSessionEntity` by those two fields.

Search by `brokerUserId` can be accomplished in two steps:

1. Try to obtain a `UserModel` by searchable attributes `UserModel.IDP_ALIAS` and `UserModel.IDP_USER_ID`
2. Search offline user sessions by the obtained `UserModel`

Search by `brokerSessionId` is more problematic though. There is no searchable `brokerSessionId` field/attribute in `UserModel`. Looking at usages from the current code base, it's used only from the back channel logout. So here is how I changed the behavior. We can first leverage `brokerUserId` from a logout token to obtain offline user sessions and then use `brokerSessionId` from the token to filter down obtained offline user sessions which will be then revoked.

 #### Enable preloading in crossDC tests
One of the CrossDC tests in the base testsuite tests whether the offline sessions were correctly preloaded and then distributed to another DC. This test requires the preloading to be enabled. However, I only managed to turn on preloading globally for all CrossDC tests. 
The scenario where the offline session preloading is turned off in CrossDC environment is covered by model tests, where I added couple of tests. 

 #### Queries for offline sessions by delegated clients from external storage
Queries by clients Id for clients delegated from external storage didn't work because we put "external" as a client Id for `PersistentClientSessionEntity`. Therefore I needed to add some logic to `JpaUserSessionPersisterProvider` and introduce new queries that search by `externalClientId` and `clientStorageProvider`.

@mposolda @stianst I would be interested in your opinions on this. Feedback from others is also welcomed. Thanks!

**Comment by mposolda** [comment]
@martin-kanis The proposal looks good to me.
